### PR TITLE
Tibor/merge all footnote

### PR DESCRIPTION
**Pull request by tiberlas**
## Description (e.g. "Related to ...", "Closes ...", etc.)

- add support for read, and insert footnotes

this is to closes issue [#4374](https://github.com/openlawlibrary/platform/issues/4374)

## Code review checklist

- [ ] Private platform tests at `core/tests/test_python-docx` are updated and passing
- [ ] If this change is going to be deployed on Cloudsmith after merge, python-docx version number should be increased 

[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: docx/parts/document.py [hunk 1]
```diff
@@ -8,7 +8,12 @@
 from itertools import chain
 from docx.document import Document
 from docx.opc.constants import RELATIONSHIP_TYPE as RT
-from docx.oxml.shape import CT_Inline
```
Comment (balsa-sarenac) [inline docx/parts/document.py:11]: Is this planned to be deleted? Is it used?

File: docx/footnotes.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            # When adding a footnote it can be inserted 
+            # in front of some other footnotes, so
+            # we need to sort footnotes by `footnote_reference_id`
+            # in |Footnotes| and in |Paragraph|
+            #
+            # resolve reference ids in |Footnotes|
+            # iterate in reverse and compare the current
+            # id with the inserted id. If there are the same
+            # insert the new footnote in that place, if not
+            # increment the current footnote id.
```
Comment (balsa-sarenac) [inline docx/footnotes.py:50]: This micro-optimization makes sense, but is depends on case-by-case basis. If user only adds footnotes at the beginning, we don't achieve any performance improvement. I'm not saying we should change it, just leaving a comment so we remember to take care of this in the future. We need some benchmarks and usage scenarios to conclude the best scenario.

File: docx/text/paragraph.py [hunk 3]
```diff
@@ -38,6 +38,47 @@ def __init__(self, p, parent):
...
+            if self._p is not paragraphs[p_i]._p:
+                if paragraphs[p_i]._p.footnote_reference_ids is not None:
+                    if not has_passed_self:
+                        for r in paragraphs[p_i].runs:
+                            r._r.increment_footnote_reference_id()
+                    else:
+                        new_fr_id = max(paragraphs[p_i]._p.footnote_reference_ids)+1
+                        break
+            else:
+                has_passed_self = True
```
Comment (balsa-sarenac) [inline docx/text/paragraph.py]: Would it make sense to move this to the document level?

**Review by balsa-sarenac — COMMENTED** [review]
Looks good; a few comments are inlined.

File: docx/parts/document.py (hunk 1 above)
Comment (tiberlas) [reply docx/parts/document.py:11]: Yes, good catch, will remove it.

File: docx/text/paragraph.py (hunk 3 above)
Comment (tiberlas) [reply docx/text/paragraph.py]: We can move bulk of this code in a new function in document level, named something like this: '_calculate_next_footnote_reference_id'.

File: docx/footnotes.py (hunk 2 above)
Comment (tiberlas) [reply docx/footnotes.py:50]: I agree, the optimization is based on premise that the draft server will add footnotes on the end of the document.

File: docx/text/paragraph.py (hunk 3 above)
Comment (balsa-sarenac) [reply docx/text/paragraph.py]: Sounds good!

File: docx/footnotes.py (hunk 2 above)
Comment (balsa-sarenac) [reply docx/footnotes.py:50]: Which is fine for now. This should be tackled in the future when contributing to upstream.

File: docx/document.py [hunk 4]
```diff
@@ -237,6 +250,39 @@ def _body(self):
...
+            if p is not self.paragraphs[p_i]._p:
+                if self.paragraphs[p_i]._p.footnote_reference_ids is not None:
+                    if not has_passed_self:
+                        for r in self.paragraphs[p_i].runs:
+                            r._r.increment_footnote_reference_id()
+                    else:
+                        new_fr_id = max(self.paragraphs[p_i]._p.footnote_reference_ids)+1
+                        break
+            else:
+                has_passed_self = True
```
Comment (balsa-sarenac) [inline docx/document.py]: Can we shuffle these ifs a bit to reduce nesting? It's very hard to reason about logic like this.
Maybe something like (not tested):

```suggestion
        for p_i in reversed(range(len(self.paragraphs))):
            # mark when we pass the paragraph containing the footnote
            if p is self.paragraphs[p_i]._p:
                has_passed_self = True
                continue
            # skip paragraphs without footnotes (they don't impact new id)
            if self.paragraphs[p_i]._p.footnote_reference_ids is None:
                continue
            # update footnote id for each passed run?
            if not has_passed_self:
                for r in self.paragraphs[p_i].runs:
                    r._r.increment_footnote_reference_id()
            else:
                # when we get to paragraph containing footnote
                # footnote id is max footnote reference id + 1?
                new_fr_id = max(self.paragraphs[p_i]._p.footnote_reference_ids)+1
                break

```

What do you think of this?

File: docx/document.py [hunk 5]
```diff
@@ -237,6 +250,39 @@ def _body(self):
...
+        # If paragraph already contains footnotes
+        # and it's the last paragraph with footnotes, then
+        # append the new footnote and the end with the next reference id.
+        if p.footnote_reference_ids is not None:
+            new_fr_id = p.footnote_reference_ids[-1] + 1
+        # If the document has footnotes after this paragraph,
+        # the increment all footnotes pass this paragraph, 
+        # and insert a new footnote at the proper position.
+        # break the loop when we get to the footnote before the one we are inserting.
+        has_passed_self = False
```
Comment (balsa-sarenac) [inline docx/document.py]: Now self is document, so this is a bit misleading in this scenario.
Maybe something like `has_passed_containing_para`?
It's a bit verbose, but it's the first thing that came to my mind.

File: docx/oxml/text/run.py [hunk 6]
```diff
@@ -73,6 +85,30 @@ def clear_content(self):
...
+        """
+        references = []
+        for child in self:
+            if child.tag == qn('w:footnoteReference'):
+               references.append(child.id)
+        if references == []:
+            references = None
+        return references
+
+    def increment_footnote_reference_id(self):
```
Comment (balsa-sarenac) [inline docx/oxml/text/run.py]: Would it make sense to rename this method to something like:
`increment_containing_footnote_references_ids`?

File: docx/document.py [hunk 7]
```diff
@@ -237,6 +250,39 @@ def _body(self):
...
+        for p_i in reversed(range(len(self.paragraphs))):
+            if p is not self.paragraphs[p_i]._p:
+                if self.paragraphs[p_i]._p.footnote_reference_ids is not None:
+                    if not has_passed_self:
+                        for r in self.paragraphs[p_i].runs:
+                            r._r.increment_footnote_reference_id()
```
Comment (balsa-sarenac) [inline docx/document.py]: For this maybe it makes sense to create method on para level that is same as run level like: `increment_continaing_footnote_reference_ids` in para so we can encapsulate this for loop in the para?

File: docx/document.py (hunk 4 above)
Comment (tiberlas) [reply docx/document.py]: Yep, it's easier to read this way, will improve it.

File: docx/document.py (hunk 5 above)
Comment (tiberlas) [reply docx/document.py]: Agree

File: docx/oxml/text/run.py (hunk 6 above)
Comment (tiberlas) [reply docx/oxml/text/run.py]: Yes, it makes sense

File: docx/document.py (hunk 7 above)
Comment (tiberlas) [reply docx/document.py]: OK, it makes sense to encapsulate it. Will improve it.

File: docx/document.py [hunk 8]
```diff
@@ -237,6 +250,46 @@ def _body(self):
...
+            # update footnote id of paragraph that is after the 
+            # paragraph that is inserting new footnote.
+            if not has_passed_containing_para:
+                self.paragraphs[p_i].increment_containing_footnote_reference_ids()
+            else:
+                # this is the first paragraph containing footnotes before the
+                # paragraph that is inserting new footnote, so we get the largest
+                # reference id and add one
+                new_fr_id = max(self.paragraphs[p_i]._p.footnote_reference_ids)+1
+                break
```
Comment (balsa-sarenac) [inline docx/document.py:290]: What if we insert a footnote in between two footnotes in a paragraph? If I followed the flow and the comments correctly it says that the footnote will have ID same as the first footnote in the paragraph? I know this is covered with tests and it should work, but I cannot understand the code flow of that case and comments didn't help as much. Can you change the second comment to focus more on why and not what the code does?

Comment (tiberlas) [reply docx/document.py:290]: Ok, will make more usefull comment.

**Review by balsa-sarenac — APPROVED** [review]

**Pull request merged**